File: oxidation/crates/parsec_api_protocol/src/message.rs
// Parsec Cloud (https://parsec.cloud) Copyright (c) BSLv1.1 (eventually AGPLv3) 2016-2021 Scille SAS

use serde::{Deserialize, Serialize};
use serde_with::{serde_as, Bytes};

use parsec_api_types::{DateTime, DeviceID};
use parsec_schema::parsec_schema;

#[parsec_schema]
pub struct MessageGetReq {
    pub offset: u64,
}

#[serde_as]
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct Message {
    pub count: u64,
    pub sender: DeviceID,
    pub timestamp: DateTime,
    #[serde_as(as = "Bytes")]
    pub body: Vec<u8>,
}

#[parsec_schema]
#[serde(tag = "status", rename_all = "snake_case")]
pub enum MessageGetRep {
    Ok { messages: Vec<Message> },
    UnknownError { error: String },
}

File: oxidation/generate_test_data/protocol/message.py
# Parsec Cloud (https://parsec.cloud) Copyright (c) BUSL-1.1 (eventually AGPL-3.0) 2016-present Scille SAS


from parsec._parsec import DateTime
from parsec.api.data import *
from parsec.api.protocol import *

from .utils import *

################### MessageGet ##################

serializer = message_get_serializer

serialized = serializer.req_dumps({"cmd": "message_get", "offset": 8})
serializer.req_loads(serialized)
display("message_get_req", serialized, [])

serialized = serializer.rep_dumps(
    {
        "messages": {
            "count": 1,
            "sender": DeviceID("alice@dev1"),
            "timestamp": DateTime(2000, 1, 2, 1),
            "body": b"foobar",
        }
    }
)
serializer.rep_loads(serialized)
display("message_get_rep", serialized, [])
